packages/babel-check-binary-expressions/src/index.js

```diff
@@ -1,6 +1,24 @@
-export default function(left, right) {
-  // eslint-disable-next-line no-undef
-  if (left instanceof BigInt && right instanceof BigInt) {
+import isNumber from "lodash/isNumber";
+import isString from "lodash/isString";
+
+function isBigInt(x) {
+  return x instanceof global.BigInt;
+}
+
+export default function(left, right, operator) {
+  // TODO: don't throw for operations that can be performed on BigInt and Number (i.e. comparisons)
+  if (
+    (isBigInt(left) && isNumber(right)) ||
+    (isNumber(left) && isBigInt(right))
+  ) {
+    throw new TypeError(`Cannot perform ${operator} on a BigInt and a Number`);
+  }
+
+  if (isString(left) || isString(right)) {
+    return left.toString() + right.toString();
```

Comment:
This is not quite right. For example, a user may monkey-patch `Number.prototype.toString`, but that behavior shouldn't be reflected here. Instead, do simply `left + right`.

In general: If you can try to follow the spec in order, one by one, you'll arrive at something closer to compliance. [Spec for +](https://tc39.github.io/proposal-bigint/#sec-addition-operator-plus-runtime-semantics-evaluation).